_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+such as a SQL database and a event bus like RabbitMQ. 
+In these distributed architectures, the ACID properties, which are the backbone of transactional integrity within a single database, 
+become challenging to enforce across multiple systems.
+
+One of the core issues is the uncertainty of atomicity. 
+For instance, consider a scenario where a transaction involves writing an event to a RabbitMQ and also updating a record in a database. 
+There's a possibility that the event write to RabbitMQ succeeds, but the outer database transaction fails. 
+This results in a scenario where an event indicating a business action is published, 
+but the corresponding database state does not reflect this action because the transaction was not completed successfully. 
+The system thus ends up in an inconsistent state, with the event bus reflecting a change that the database does not.
+
+When integrating an event bus or any other external service within an application in a distributed system, there are two naive approaches one might consider. 
+Each approach has its own set of challenges and potential pitfalls:
+ 
+## 1. Execute Business Logic Within the Transaction and Publish Events After Committing the Transaction
+In this approach, the application first executes all the required business logic within a database transaction. 
+After this transaction is successfully committed, it then publishes the relevant events to the event bus.
+
+![commit-publish](/assets/article_images/commit-publish.png)
+
+**Challenges:**
+
+* **Risk of Event Loss:** If the application crashes or encounters an issue between the transaction commit and the event publication, the event may never get published. 
+This leads to a situation where the database state has changed, but the corresponding event informing other services of this change is lost.
+* **Inconsistency in Distributed Systems**: Other services relying on these events for triggering their own processes will not be aware of the changes, leading to data inconsistency across the system.
+* **Complex Recovery Mechanisms**: Implementing a mechanism to check for such failures and recover from them can be complex and error-prone.
+
+This approach can be suitable for situations where the events being published are not critical to the core business logic or data integrity of the application. 
+An example of this might be events used for monitoring or logging purposes. 
+In these cases, the loss of some events due to system failures or crashes might be considered an acceptable risk.
+
+## 2. Execute Business Logic and Publish Events Within the Transaction
+In this approach, the application attempts to execute business logic and publish events to the event bus, all within the same database transaction. 
+
+
+![publish-commit](/assets/article_images/publish-commit.png)
+
+**Challenges:**
```

Comment:
I'm not an expert on transaction commits, maybe provide a link for further reading on the word "commit"?